File: scripts/check-nav-v2-operational-duplicate-review.mjs
import assert from 'node:assert/strict';
import {
  buildExactDuplicateOwnerDecisionPackage,
  createExactDuplicateDecisionState,
  summarizeExactDuplicateOwnerDecision,
  updateExactDuplicateDecisionState,
  validateExactDuplicateReviewReport
} from '../assets/js/nav-v2/operational-duplicate-review-model-v2.js';

function deal(id, createdAt, overrides = {}) {
  return {
    deal_id: id,
    deal_title: `Сделка ${id}`,
    address: 'Тестовый адрес',
    status: 'draft',
    risk_level: 'yellow',
    readiness_deposit: 80,
    readiness_deal: 50,
    created_at: createdAt,
    updated_at: createdAt,
    latest_activity_at: createdAt,
    next_action: 'Проверить карточку',
    counts: { tasks: 2, risks: 1, documents: 8, events: 2, comments: 0, reviews: 0, participants: 1, expenses: 2 },
    latest: {},
    semantic_hashes: { deal: 'a', tasks: 'b', risks: 'c', documents: 'd', events: 'e', comments: 'f', reviews: 'f', participants: 'g', expenses: 'h' },
    card_url: `./deal-card-v2.html?id=${id}`,
    ...overrides
  };
}

function group(key, firstId, secondId, overrides = {}) {
  return {
    group_key: key,
    created_by: 'owner-1',
    created_by_name: 'Алексей Ковтун',
    deal_count: 2,
    first_created_at: '2026-07-14T09:00:00.000Z',
    last_created_at: '2026-07-14T09:00:06.000Z',
    interval_seconds: 6,
    suggested_canonical_deal_id: firstId,
    suggestion_basis: 'earliest_created_only',
    suggestion_confidence: 'medium',
    all_semantic_equal: true,
    has_post_creation_divergence: false,
    entity_comparison: { deal: true, tasks: true, risks: true, documents: true, events: true, comments: true, reviews: true, participants: true, expenses: true },
    comments_and_reviews: 0,
    manual_review_reasons: ['Текущие карточки совпадают.', 'Раннейшая карточка — только предложение.'],
    deals: [deal(firstId, '2026-07-14T09:00:00.000Z'), deal(secondId, '2026-07-14T09:00:06.000Z')],
    selection_available: false,
    mutation_available: false,
    owner_decision_required: true,
    ...overrides
  };
}

function report(role = 'owner') {
  const items = [
    group('group-1', 'deal-1', 'deal-2'),
    group('group-2', 'deal-3', 'deal-4', {
      all_semantic_equal: false,
      has_post_creation_divergence: true,
      suggestion_confidence: 'low',
      entity_comparison: { deal: true, tasks: false, risks: true, documents: true, events: true, comments: true, reviews: true, participants: true, expenses: true },
      manual_review_reasons: ['Различаются задачи.', 'Нужен перенос уникальных данных.']
    })
  ];
  return {
    report_version: 8,
    profile: { id: `${role}-1`, role, full_name: role === 'manager' ? 'Менеджер' : 'Владелец', email: `${role}@example.test` },
    exact_duplicate_review_pack: {
      review_version: 1,
      generated_at: '2026-07-14T12:00:00.000Z',
      summary: {
        groups: 2,
        deals: 4,
        exact_semantic_groups: 1,
        diverged_groups: 1,
        groups_with_comments_or_reviews: 0,
        selection_available: false,
        mutation_available: false,
        cleanup_execution_available: false,
        owner_decision_required: true
      },
      items
    }
  };
}

const validation = validateExactDuplicateReviewReport(report());
assert.equal(validation.valid, true);
assert.equal(validation.actor_can_decide, true);
assert.equal(validation.groups.length, 2);

let state = createExactDuplicateDecisionState(validation);
state = updateExactDuplicateDecisionState(state, 'group-1', {
  decision_status: 'confirmed',
  canonical_deal_id: 'deal-1',
  resolution: 'archive_duplicate',
  transfer_note: 'Карточки полностью совпадают, перенос уникальных данных не требуется.',
  decision_reason: 'Ранняя карточка выбрана после сравнения всех дочерних сущностей.'
});
state = updateExactDuplicateDecisionState(state, 'group-2', {
  decision_status: 'confirmed',
  canonical_deal_id: 'deal-3',
  resolution: 'merge_then_archive',
  transfer_note: 'Перенести уникальную задачу из второй карточки перед архивированием.',
  decision_reason: 'В первой карточке сохранён исходный контур, во второй есть уникальная задача.'
});

const summary = summarizeExactDuplicateOwnerDecision(validation, state, report().profile);
assert.equal(summary.decision_package_ready, true);
assert.equal(summary.confirmed, 2);
assert.equal(summary.cleanup_candidate_groups, 2);
assert.equal(summary.cleanup_authorized, false);

const payload = buildExactDuplicateOwnerDecisionPackage(validation, state, report().profile, {
  generatedAt: '2026-07-14T12:10:00.000Z',
  reviewGeneratedAt: '2026-07-14T12:00:00.000Z'
});
assert.equal(payload.summary.decision_package_ready, true);
assert.equal(payload.summary.cleanup_authorized, false);
assert.equal(payload.safety.server_mutation_available, false);
assert.equal(payload.safety.automatic_canonical_selection_available, false);
assert.equal(payload.safety.requires_one_group_at_a_time, true);

const managerValidation = validateExactDuplicateReviewReport(report('manager'));
const managerSummary = summarizeExactDuplicateOwnerDecision(managerValidation, createExactDuplicateDecisionState(managerValidation), report('manager').profile);
assert.equal(managerValidation.valid, true);
assert.equal(managerValidation.actor_can_decide, false);
assert.equal(managerSummary.decision_package_ready, false);

const invalidCanonical = updateExactDuplicateDecisionState(state, 'group-1', { canonical_deal_id: 'deal-other' });
const invalidCanonicalSummary = summarizeExactDuplicateOwnerDecision(validation, invalidCanonical, report().profile);
assert.equal(invalidCanonicalSummary.decision_package_ready, false);
assert.match(invalidCanonicalSummary.decision_rows[0].errors.join(' '), /каноническую/i);

const missingTransfer = updateExactDuplicateDecisionState(state, 'group-2', { transfer_note: '' });
const missingTransferSummary = summarizeExactDuplicateOwnerDecision(validation, missingTransfer, report().profile);
assert.equal(missingTransferSummary.decision_package_ready, false);
assert.match(missingTransferSummary.decision_rows[1].errors.join(' '), /переноса/i);

let needsReviewState = createExactDuplicateDecisionState(validation);
for (const key of ['group-1', 'group-2']) {
  needsReviewState = updateExactDuplicateDecisionState(needsReviewState, key, {
    decision_status: 'needs_review',
    resolution: 'needs_manual_review',
    decision_reason: 'Нужно дополнительное подтверждение владельца и сотрудников по карточке.'
  });
}
const needsReviewSummary = summarizeExactDuplicateOwnerDecision(validation, needsReviewState, report().profile);
assert.equal(needsReviewSummary.decision_package_ready, true);
assert.equal(needsReviewSummary.needs_review, 2);
assert.equal(needsReviewSummary.cleanup_candidate_groups, 0);

const tamperedReport = report();
tamperedReport.exact_duplicate_review_pack.summary.mutation_available = true;
const tamperedValidation = validateExactDuplicateReviewReport(tamperedReport);
assert.equal(tamperedValidation.valid, false);
assert.match(tamperedValidation.errors.join(' '), /mutation_available/);

const duplicateGroupReport = report();
duplicateGroupReport.exact_duplicate_review_pack.items[1].group_key = 'group-1';
const duplicateGroupValidation = validateExactDuplicateReviewReport(duplicateGroupReport);
assert.equal(duplicateGroupValidation.valid, false);
assert.match(duplicateGroupValidation.errors.join(' '), /повторяющиеся group_key/i);

const duplicateDealReport = report();
duplicateDealReport.exact_duplicate_review_pack.items[1].deals[0].deal_id = 'deal-1';
const duplicateDealValidation = validateExactDuplicateReviewReport(duplicateDealReport);
assert.equal(duplicateDealValidation.valid, false);
assert.match(duplicateDealValidation.errors.join(' '), /нескольких группах/i);

console.log('Navigator v2 exact duplicate review semantic regression passed');

File: scripts/check_nav_v2_operational_duplicate_review.py
from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
MIGRATION = ROOT / "supabase/migrations/20260714130000_nav_v2_exact_duplicate_review_pack.sql"
PAGE = ROOT / "operational-duplicate-review-v2.html"
UI = ROOT / "assets/js/nav-v2/operational-duplicate-review-v2.js"
MODEL = ROOT / "assets/js/nav-v2/operational-duplicate-review-model-v2.js"
SEMANTIC = ROOT / "scripts/check-nav-v2-operational-duplicate-review.mjs"
SOURCE_PAGE = ROOT / "operational-adoption-v2.html"
REGISTRY = ROOT / "config/nav-v2-rpc-surface.json"
BUDGET = ROOT / "config/nav-v2-module-budget.json"
STATIC_WORKFLOW = ROOT / ".github/workflows/nav-v2-static.yml"
DEDICATED_WORKFLOW = ROOT / ".github/workflows/nav-v2-operational-duplicate-review.yml"
PUBLIC_SMOKE = ROOT / "tests/e2e/public-smoke.spec.js"


def require(text: str, markers: tuple[str, ...], label: str, errors: list[str]) -> None:
    for marker in markers:
        if marker not in text:
            errors.append(f"{label}: missing {marker!r}")


def assert_read_only(text: str, label: str, errors: list[str]) -> None:
    lowered = text.lower()
    for forbidden in (
        "update public.nav_deals_v2",
        "insert into public.nav_deals_v2",
        "delete from public.nav_deals_v2",
        "update public.nav_deal_tasks_v2",
        "insert into public.nav_deal_tasks_v2",
        "delete from public.nav_deal_tasks_v2",
        "update public.nav_deal_risks_v2",
        "insert into public.nav_deal_risks_v2",
        "delete from public.nav_deal_risks_v2",
        "update public.nav_deal_documents_v2",
        "insert into public.nav_deal_documents_v2",
        "delete from public.nav_deal_documents_v2",
    ):
        if forbidden in lowered:
            errors.append(f"{label} must remain read-only: {forbidden}")


def main() -> int:
    errors: list[str] = []
    for path in (
        MIGRATION, PAGE, UI, MODEL, SEMANTIC, SOURCE_PAGE,
        REGISTRY, BUDGET, STATIC_WORKFLOW, DEDICATED_WORKFLOW, PUBLIC_SMOKE,
    ):
        if not path.exists():
            errors.append(f"missing {path.relative_to(ROOT)}")
    if errors:
        print("\n".join(errors))
        return 1

    sql = MIGRATION.read_text(encoding="utf-8")
    require(sql, (
        "nav_v2_private.nav_v2_get_exact_duplicate_review_pack_unchecked_20260714(",
        "security definer",
        "set search_path = public, pg_temp",
        "v_role is null or v_role not in ('owner', 'admin', 'manager')",
        "coalesce(deal.title, '') not ilike 'ДЕМО:%'",
        "array['deal_id','task_id','document_id','risk_id','review_id']",
        "all_semantic_equal",
        "suggested_canonical_deal_id",
        "'suggestion_basis', 'earliest_created_only'",
        "'selection_available', false",
        "'mutation_available', false",
        "'cleanup_execution_available', false",
        "'owner_decision_required', true",
        "'report_version', 8",
        "'exact_duplicate_review_pack', v_duplicate_review",
        "revoke execute on function nav_v2_private.nav_v2_get_exact_duplicate_review_pack_unchecked_20260714(integer) from anon, authenticated",
        "grant execute on function nav_v2_private.nav_v2_get_exact_duplicate_review_pack_unchecked_20260714(integer) to service_role",
        "Exact duplicate review implementation must remain private",
    ), MIGRATION.name, errors)
    assert_read_only(sql, MIGRATION.name, errors)

    page = PAGE.read_text(encoding="utf-8")
    require(page, (
        "Content-Security-Policy",
        'aria-live="polite"',
        "assets/js/nav-v2/operational-duplicate-review-v2.js?v=20260714-01",
        "assets/js/nav-v2/role-menu-v2.js?v=20260713-01",
        "assets/js/nav-v2/nav-base-menu-cleanup-v2.js?v=20260707-1210",
    ), PAGE.name, errors)

    ui = UI.read_text(encoding="utf-8")
    require(ui, (
        "validateExactDuplicateReviewReport",
        "createExactDuplicateDecisionState",
        "summarizeExactDuplicateOwnerDecision",
        "buildExactDuplicateOwnerDecisionPackage",
        "['owner', 'admin', 'manager'].includes(profile().role)",
        "['owner', 'admin'].includes(profile().role)",
        "exact_duplicate_review_pack",
        "navigator-v2-exact-duplicate-owner-decision-",
        "rpc('nav_v2_get_operational_adoption_report'",
        "Рекомендация не является выбором",
        "cleanup_authorized=false",
        "URL.createObjectURL",
    ), UI.name, errors)
    if ui.count("rpc('nav_v2_get_operational_adoption_report'") != 1:
        errors.append("duplicate review UI must use exactly one adoption report RPC call")
    for forbidden in (
        "nav_v2_update_",
        "nav_v2_add_",
        "nav_v2_save_",
        ".from('nav_",
        '.from("nav_',
        "localStorage",
        "sessionStorage",
    ):
        if forbidden in ui:
            errors.append(f"duplicate review UI must remain browser-local and read-only: {forbidden}")

    model = MODEL.read_text(encoding="utf-8")
    require(model, (
        "export function validateExactDuplicateReviewReport",
        "export function duplicateResolutionOptions",
        "export function createExactDuplicateDecisionState",
        "export function updateExactDuplicateDecisionState",
        "export function summarizeExactDuplicateOwnerDecision",
        "export function buildExactDuplicateOwnerDecisionPackage",
        "navigator_v2_exact_duplicate_review_validation",
        "navigator_v2_exact_duplicate_owner_decision",
        "decision_package_ready",
        "cleanup_authorized: false",
        "server_mutation_available: false",
        "automatic_canonical_selection_available: false",
        "automatic_merge_available: false",
        "automatic_archive_available: false",
        "requires_fresh_server_revalidation: true",
        "requires_audit_event: true",
        "requires_one_group_at_a_time: true",
    ), MODEL.name, errors)
    for forbidden in ("rpc(", ".from(", "fetch(", "localStorage", "sessionStorage"):
        if forbidden in model:
            errors.append(f"duplicate review model must remain pure and offline: {forbidden}")

    semantic = SEMANTIC.read_text(encoding="utf-8")
    require(semantic, (
        "decision_package_ready, true",
        "cleanup_authorized, false",
        "managerValidation",
        "invalidCanonical",
        "missingTransfer",
        "needsReviewState",
        "tamperedValidation",
        "duplicateGroupValidation",
        "duplicateDealValidation",
        "semantic regression passed",
    ), SEMANTIC.name, errors)

    source_page = SOURCE_PAGE.read_text(encoding="utf-8")
    if "operational-duplicate-review-v2.html" not in source_page:
        errors.append("operational adoption page must link to duplicate review")

    registry = json.loads(REGISTRY.read_text(encoding="utf-8"))
    helper = "nav_v2_get_exact_duplicate_review_pack_unchecked_20260714"
    if registry.get("internal_only", []).count(helper) != 1:
        errors.append("duplicate review helper must be registered exactly once as internal_only")
    for category in ("frontend_api", "admin_api", "demo_api"):
        if helper in registry.get(category, []):
            errors.append(f"duplicate review helper leaked into {category}")

    budget = json.loads(BUDGET.read_text(encoding="utf-8"))
    if (budget.get("pages") or {}).get(PAGE.name) != {"max_modules": 3}:
        errors.append("duplicate review page must have a three-module budget")

    static_workflow = STATIC_WORKFLOW.read_text(encoding="utf-8")
    dedicated_workflow = DEDICATED_WORKFLOW.read_text(encoding="utf-8")
    python_command = "python3 scripts/check_nav_v2_operational_duplicate_review.py"
    node_command = "node scripts/check-nav-v2-operational-duplicate-review.mjs"
    for workflow, label in ((static_workflow, STATIC_WORKFLOW.name), (dedicated_workflow, DEDICATED_WORKFLOW.name)):
        if python_command not in workflow:
            errors.append(f"{label}: missing duplicate review static regression")
        if node_command not in workflow:
            errors.append(f"{label}: missing duplicate review semantic regression")

    public_smoke = PUBLIC_SMOKE.read_text(encoding="utf-8")
    if "'/operational-duplicate-review-v2.html'" not in public_smoke:
        errors.append("public smoke must include operational-duplicate-review-v2.html guest gate")

    if errors:
        print("Navigator v2 exact duplicate review errors:")
        for error in errors:
            print(f"- {error}")
        return 1

    print(
        "Navigator v2 exact duplicate review passed: demo exclusion, normalized child comparisons, private helper, "
        "manager read-only view, owner/admin local decision package, one browser RPC and no cleanup mutation"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
